### DISCUSSION TITLE
Merge task failed

### DISCUSSION THREAD

**Discussion by ariel234156** [nextpyp/.github#111] [Q&A] [answered]
hello,
I did particle refinement, and the iterations ran well, but in the merge task, the running failed. I have attached the code failure in the pictures-
![Screenshot (241)](https://github.com/nextpyp/.github/assets/137890314/d5311e95-09ad-4862-be17-87c1960d8bfa)
![Screenshot (244)](https://github.com/nextpyp/.github/assets/137890314/1c10d1b0-1821-4348-80e7-7902a4790727)
![Screenshot (243)](https://github.com/nextpyp/.github/assets/137890314/b56e63f8-1901-49f1-b339-45d776f244d0)
![Screenshot (242)](https://github.com/nextpyp/.github/assets/137890314/b2c01495-964d-4c40-b8f9-319301fefc7a)

Someone knew what the problem was?
Thank you.

**Comment by abartesaghi** [comment]
How many particles do you have in these tilt-series? The logs seem to indicate that no particles were actually extracted.

**Comment by ariel234156** [comment]
Thank you for your reply,
The software extracted 90000 particles in preprocces step, but I don't know
if this extraction is real.

**Reply by abartesaghi** [reply]
OK, do you see any blue circles in the Reconstruction tab (pre-processing block)? And, where is the 90k number reported?

**Comment by ariel234156** [comment]
Yes, I see the blue circles.
The 90k particles report in the end of particles detection before particles
renfiment.

**Reply by abartesaghi** [reply]
Can you share the contents of the `Commands` field at the top the logs for the corresponding Launch job?

**Comment by ariel234156** [comment]
Hi Alberto,
Sorry for the late response,
Because this step failed, I removed it. I will rerun it and send you the commands.
Thank you.

Ariel Cohen
MSc student in Physical Chemistry at Raviv lab
Los Angeles building, room 23.
Phone Number- +972523969740
[cid:231f8988-66ba-4216-98f5-513cc72a2c78]
________________________________

**Comment by ariel234156** [comment]
Hi Alberto, i attached the commands from particle refinement process:

"/opt/pyp/bin/run/csp" -data_parent "/mnt/nvme1n1/users/coari/projects/ITT-KABY/tomo-preprocessing-fAqQeIsEqhk6MI0Z" -particle_mw 55.0 -particle_rad 50.0 -particle_sym "C14" -extract_box 256 -extract_bin 2 -extract_fmt frealign -refine_model "/mnt/nvme1n1/users/coari/YK_20240402_154016/outputfile.mrc" -refine_parfile_tomo "/mnt/nvme1n1/users/coari/projects/ITT-KABY/tomo-preprocessing-fAqQeIsEqhk6MI0Z/frealign/20240521_180038_tomo-preprocessing-fAqQeIsEqhk6MI0Z_original_volumes.txt" -refine_rhref "8" -refine_fboost -refine_skip -reconstruct_maxtilt 60.0 -reconstruct_mintilt -60.0 -csp_UseImagesForRefinementMax 10 -csp_NumberOfRandomIterations 50000 -csp_refine_particles -csp_ToleranceParticlesPhi 10.0 -csp_ToleranceParticlesTheta 10.0 -csp_ToleranceParticlesShifts 50.0 -slurm_tasks 50 -slurm_memory 400 -slurm_walltime "9:00:00" -slurm_merge_tasks 50 -slurm_merge_memory 400 -slurm_merge_walltime "9:00:00"

Thank you.

Ariel Cohen
MSc student in Physical Chemistry at Raviv lab
Los Angeles building, room 23.
Phone Number- +972523969740
[cid:92b226a0-04f5-4de7-9c49-72ff1bd17209]
________________________________

**Comment by abartesaghi** [comment]
The particles are definitely not being extracted for some reason. Here are some things to check:
- Can you confirm that the following file is not empty:
`/mnt/nvme1n1/users/coari/projects/ITT-KABY/tomo-preprocessing-fAqQeIsEqhk6MI0Z/frealign/20240521_180038_tomo-preprocessing-fAqQeIsEqhk6MI0Z_original_volumes.txt`?
- Try re-running the refinement block in verbose mode by checking the `Verbose mode` box in the Resources tab and seeing if there are any other warning messages.
- It looks like you are using geometry-based picking, is the 90k number reported inside the pre-processing block after you run particle picking?
- Also, I assume the blue circles are showing up in the right places in 3D, can you confirm this?

**Comment by ariel234156** [comment]
Hi Alberto,
Thank you for the quick response.

  1.
The file that you mention isn't empty.
  2.
I will do "verbose" and send you the logs of error code.
  3.
I saw the blue circle in 3D after particle picking in the preprocessing in the right place.

Because the 90k particle creates 7Tb of data, I tried to use it in the smaller dataset, and I got the same problem.


Ariel Cohen
MSc student in Physical Chemistry at Raviv lab
Los Angeles building, room 23.
Phone Number- +972523969740
[cid:07ed4f9f-1b6f-45e2-8b79-5f8e0f4a87d8]
________________________________

**Reply by ariel234156** [reply]
I attached the log from with verbose option.
![Screenshot (247)](https://github.com/nextpyp/.github/assets/137890314/ce49d616-cfb7-4ec4-a4d4-209f3854d9ad)
![Screenshot (246)](https://github.com/nextpyp/.github/assets/137890314/3554dbe5-c9aa-4e0a-ace0-4e3b0bc3b4ea)
![Screenshot (245)](https://github.com/nextpyp/.github/assets/137890314/74b839c4-a801-468a-9142-93816233dcca)

**Reply by abartesaghi** [reply]
This log shows that there were 11 particle projections extracted, which is different from before. Also, since this is a re-run, it just retrieved the metadata from the previous run so the logs won't show if there was a problem during extraction. You will need to clear the block by selecting `Delete files and data` and re-run again to get the proper logs. It will also be useful to know exactly how many particles are in this particular tilt-series which you can get from the Reconstruction tab in the pre-processing block.

**Comment by ariel234156** [comment]
Hi,
I did the preprocess and particle refinement again from the start and got 135000 particles in the preprocessing step. I got the error again in the refinement step. I attached the command for these steps and photos of all the logs.
commands from particle refinement-
"/opt/pyp/bin/run/csp" -data_parent "/mnt/nvme1n1/users/coari/projects/ITT-KABY/tomo-preprocessing-fB7wGmYvJgjdi68y" -particle_mw 55.0 -particle_rad 50.0 -particle_sym "C14" -extract_box 512 -extract_fmt frealign -refine_model "/mnt/nvme1n1/users/coari/YK_20240402_154016/Refine3D/job440/run_half1_class001_unfil.mrc" -csp_refine_micrographs -csp_refine_particles -csp_ToleranceParticlesShifts 50.0 -slurm_verbose -slurm_tasks 50 -slurm_memory 400 -slurm_walltime "9:00:00" -slurm_merge_tasks 50 -slurm_merge_memory 400 -slurm_merge_walltime "9:00:00"

/bin/bash -c "cd '/mnt/nvme1n1/users/coari/projects/ITT-KABY/tomo-coarse-refinement-fBCtfXeuXI7Kei2l/swarm'; export cspswarm=cspswarm; /opt/pyp/bin/run/pyp --file Position_12_2 --iter 2 --no-skip --no-debug 2>&1 | tee ../log/Position_12_2_csp.log"

![Screenshot (255)](https://github.com/nextpyp/.github/assets/137890314/3c88cb48-90a8-4039-a398-8847fbd5b892)
![Screenshot (254)](https://github.com/nextpyp/.github/assets/137890314/b800ba25-669d-4b46-b02a-dc809598705e)
![Screenshot (253)](https://github.com/nextpyp/.github/assets/137890314/a92f723d-e4cd-4020-a34c-59b6b90934ea)
![Screenshot (252)](https://github.com/nextpyp/.github/assets/137890314/09229bae-11d1-4205-99c0-cea8208b5133)
![Screenshot (251)](https://github.com/nextpyp/.github/assets/137890314/3455ec15-ffbc-4f33-bfb5-27c34c51964e)
![Screenshot (250)](https://github.com/nextpyp/.github/assets/137890314/7961d485-f53b-4c05-9ad4-98018f624979)
![Screenshot (249)](https://github.com/nextpyp/.github/assets/137890314/c35182b9-5306-4c68-a0e1-3bc7e113d569)
![Screenshot (248)](https://github.com/nextpyp/.github/assets/137890314/c22fb774-a9d4-4ac8-b61f-897f9ddf9fd5)

finally i add picture of the project-
![Screenshot (257)](https://github.com/nextpyp/.github/assets/137890314/ffddf7ec-2d78-4b73-b70d-6b1efd778b80)
![Screenshot (256)](https://github.com/nextpyp/.github/assets/137890314/a96f7b1a-7a7f-402f-9ec1-f3e6f52f836e)

**Reply by abartesaghi** [reply]
It seems to be working now! You just need to increase the `Walltime (merge task)` in the Resources tab because the merge job is timing out. The walltime needs to be long enough to allow all tilt-series to finish processing. If you are not sure how long that is and/or if it doesn't make a difference in your setup, you can just use a large number.

**Comment by ariel234156** [comment]
I tried to run the command in 48 hours, but I got this error again during the split part. So, I don't think the problem is the timing.
In addition, the system wrote on all of these steps successfully despite this error(I attached an example).
![Screenshot (259)](https://github.com/nextpyp/.github/assets/137890314/201b68f7-424a-4e57-85bd-63f1996a0c72)
`<html>
<body>



ERROR: pyp exiting at 2024-05-27 16:02:36
--
226 | Traceback (most recent call last):
227 | File "/opt/pyp/bin/run/pyp", line 3949, in <module>
228 | particle_cspt.merge_movie_files_in_job_arr(
229 | File "/opt/pyp/src/pyp/utils/timer.py", line 78, in wrapper_timer
230 | return func(*args, **kwargs)
231 | File "/opt/pyp/src/pyp/refine/csp/particle_cspt.py", line 547, in merge_movie_files_in_job_arr
232 | run_reconstruction(
233 | File "/opt/pyp/src/pyp/utils/timer.py", line 78, in wrapper_timer
234 | return func(*args, **kwargs)
235 | File "/opt/pyp/src/pyp/refine/csp/particle_cspt.py", line 789, in run_reconstruction
236 | par_obj = call_shape_phase_residuals(
237 | File "/opt/pyp/src/pyp/utils/timer.py", line 78, in wrapper_timer
238 | return func(*args, **kwargs)
239 | File "/opt/pyp/src/pyp/analysis/scores.py", line 824, in call_shape_phase_residuals
240 | par_obj = shape_phase_residuals(
241 | File "/opt/pyp/src/pyp/analysis/scores.py", line 436, in shape_phase_residuals
242 | thresholds[g, f] = np.sort(meanscore)[
243 | IndexError: index 0 is out of bounds for axis 0 with size 0
244 |  
245 | During handling of the above exception, another exception occurred:
246 |  
247 | Traceback (most recent call last):
248 | File "/opt/pyp/bin/run/pyp", line 3969, in <module>
249 | trackback()
250 | File "/opt/pyp/bin/run/pyp", line 3164, in trackback
251 | get_free_space(Path(os.environ["PYP_SCRATCH"]).parents[0])
252 | File "/usr/local/envs/pyp/lib/python3.8/pathlib.py", line 621, in __getitem__
253 | raise IndexError(idx)
254 | IndexError: 0
255 |  
256 | During handling of the above exception, another exception occurred:
257 |  
258 | Traceback (most recent call last):
259 | File "/opt/pyp/bin/run/pyp", line 4591, in <module>
260 | trackback()
261 | File "/opt/pyp/bin/run/pyp", line 3164, in trackback
262 | get_free_space(Path(os.environ["PYP_SCRATCH"]).parents[0])
263 | File "/usr/local/envs/pyp/lib/python3.8/pathlib.py", line 621, in __getitem__
264 | raise IndexError(idx)
265 | IndexError: 0


</body>
</html>`

Thank you, I really appreciate your help.

**Reply by abartesaghi** [reply, marked as answer]
This looks like an issue we saw before where multiple split jobs can interfere with each other, essentially deleting each other's files and causing errors. There is an advanced parameter in the Resources tab called `Timeout to delete zombie runs (minutes)` which defaults to 600. If your runs are longer than that, you could increase that number and see if the problem goes away. Another thing to keep in mind is that you need to make sure the compute nodes have enough scratch space to save the necessary particle stacks. And if you have multiple tilt-series running on the same node at the same time, you will need to add their storage requirements and ensure that the node has enough capacity.

**Comment by ariel234156** [comment]
I changed the parameters as you suggested and now it's working, thank you for all your help @abartesaghi
